### ISSUE TITLE
I tried to port this to DriverKit

### ISSUE THREAD

**Issue by noah-nuebling** [360Controller/360Controller#1267]
Hey guys, I tried to port this driver to DriverKit, so it works under modern macOS. 

Here's the repo for my port: https://github.com/noah-nuebling/mac-gamepad-driver

Unfortunately it doesn't quite work so far but I thought I'd share it in case somebody else wants to keep working on it.

## Discussion

This attempts to port the driver for wired controllers in the original repo to DriverKit. In the original repo, the driver for wired controllers is implemented by the files `_60Controller.cpp` and `Controller.cpp`.

The port creates two clients for the gamepad in the IORegistry. `gamepad_driver_user` and `Xbox360ControllerClass`. Here's a screenshot.
<img width="1671" alt="CleanShot 2023-09-02 at 23 06 51@2x" src="https://github.com/360Controller/360Controller/assets/40808343/e34cfe26-b4b0-4caf-ab07-1c8c0137421b">

Here's more details on how it works:

I have an XboxOne wired controller. When I attach it to the computer, macOS creates a registry entry of class `IOUSBHostDevice` for it (it does this anyways even if this driver is not installed). Then, if this driver is installed, the system creates the `gamepad_driver_user` registry entry which is driven by the code in the cpp class `gamepad_driver_user` and attaches it to the `IOUSBHostDevice` as a client. When  `gamepad_driver_user` starts, it starts finding USB interfaces and stuff of the device and it tries to determine what kind of gamepad is being used (XboxOne, Xbox360, XboxOriginal, and some other options).  It tries to find the USB endpoints on the device for sending and receiving data. Then it writes some data to the device. It asks the system to create another registry entry which is driven by the code in the cpp class `Xbox360ControllerClass` and the system attaches this new entry to `gamepad_driver_user` as a client. Then `gamepad_driver_user` starts reading from the device in a loop. The data it reads from the device in a loop is sent to `Xbox360ControllerClass→handleReport()`, which then does some additional processing on the report before passing it on to userspace clients. 

The `Xbox360ControllerClass` registry entry is a subclass of `IOHIDDevice`. Which means userspace programs can find it and start talking to it using the `IOHIDManager` apis.  This architecture is the same as the original 360Controller driver. But the way it's implemented is pretty different because all the APIs changed a lot for DriverKit. 

My current implementation doesn't work properly for my wired XboxOne controller because all of the reports that are being read from the device don't seem to be "valid" according to the logic I copied from the 360Controller driver. I don't know if my controller is weird or if there's a bug in my code.

Also see `Notes.md` in the repo for more info and steps to run and test the driver.

Another thing that's really annoying is that you need to request permissions from Apple to attach your driver directly to the `IOUSBHostDevice`. And you need to request separate permissions for each vendorID. This might make it infeasible to ever publish the DriverKit based driver. But currently I'm using Ad Hoc signing and some other hacks to get around this requirement and run the driver even without permission from Apple. I even requested the permissions from Apple but they declined :(  

Another side note: 

Here are the logs that the driver outputs on my system:
<img width="734" alt="CleanShot 2023-09-02 at 23 27 18@2x" src="https://github.com/360Controller/360Controller/assets/40808343/5f8187f7-93b0-45ad-ad82-02f068d1c2b1">

I think next steps for further developing the driver would be: 
- Testing and debugging with a wired Xbox 360 controller and getting it work in the first place.
- Implementing the proper support for XboxOne and XboxOriginal controllers (stubs for XboxOne controller support are already in the repo)
- Maybe building a UI / updating the UI over from the original 360Controller driver. The old UI should work the same with the new driver without any meaningful updates I think. But it might be a good idea to port the UI over from being a prefpane to being a normal app, because prefpanes are really hard to develop under modern macOS.
- Maybe add the remapping and deadzone customization features like the original 360Controller driver, so everything in the UI works. Shouldn't be too much work.
- Implementing support for peripherals like the microphone if anyone wants to.
- Fixing all the memory leaks and bad error handling in my horrible code

If you have any more questions, let me know.

**Comment by Coresi7** [comment]
And you, my friend, are a true hero.

**Comment by jonasw8** [comment]
> Hey guys, I tried to port this driver to DriverKit, so it works under modern macOS.
> 
> Here's the repo for my port: https://github.com/noah-nuebling/mac-gamepad-driver
> 
> Unfortunately it doesn't quite work so far but I thought I'd share it in case somebody else wants to keep working on it.
> ## Discussion
> 
> This attempts to port the driver for wired controllers in the original repo to DriverKit. In the original repo, the driver for wired controllers is implemented by the files `_60Controller.cpp` and `Controller.cpp`.
> 
> The port creates two clients for the gamepad in the IORegistry. `gamepad_driver_user` and `Xbox360ControllerClass`. Here's a screenshot. <img alt="CleanShot 2023-09-02 at 23 06 51@2x" width="1671" src="https://user-images.githubusercontent.com/40808343/265231412-e34cfe26-b4b0-4caf-ab07-1c8c0137421b.png">
> 
> Here's more details on how it works:
> 
> I have an XboxOne wired controller. When I attach it to the computer, macOS creates a registry entry of class `IOUSBHostDevice` for it (it does this anyways even if this driver is not installed). Then, if this driver is installed, the system creates the `gamepad_driver_user` registry entry which is driven by the code in the cpp class `gamepad_driver_user` and attaches it to the `IOUSBHostDevice` as a client. When `gamepad_driver_user` starts, it starts finding USB interfaces and stuff of the device and it tries to determine what kind of gamepad is being used (XboxOne, Xbox360, XboxOriginal, and some other options). It tries to find the USB endpoints on the device for sending and receiving data. Then it writes some data to the device. It asks the system to create another registry entry which is driven by the code in the cpp class `Xbox360ControllerClass` and the system attaches this new entry to `gamepad_driver_user` as a client. Then `gamepad_driver_user` starts reading from the device in a loop. The data it reads from the device in a loop is sent to `Xbox360ControllerClass→handleReport()`, which then does some additional processing on the report before passing it on to userspace clients.
> 
> The `Xbox360ControllerClass` registry entry is a subclass of `IOHIDDevice`. Which means userspace programs can find it and start talking to it using the `IOHIDManager` apis. This architecture is the same as the original 360Controller driver. But the way it's implemented is pretty different because all the APIs changed a lot for DriverKit.
> 
> My current implementation doesn't work properly for my wired XboxOne controller because all of the reports that are being read from the device don't seem to be "valid" according to the logic I copied from the 360Controller driver. I don't know if my controller is weird or if there's a bug in my code.
> 
> Also see `Notes.md` in the repo for more info and steps to run and test the driver.
> 
> Another thing that's really annoying is that you need to request permissions from Apple to attach your driver directly to the `IOUSBHostDevice`. And you need to request separate permissions for each vendorID. This might make it infeasible to ever publish the DriverKit based driver. But currently I'm using Ad Hoc signing and some other hacks to get around this requirement and run the driver even without permission from Apple. I even requested the permissions from Apple but they declined :(
> 
> Another side note:
> 
> Here are the logs that the driver outputs on my system: <img alt="CleanShot 2023-09-02 at 23 27 18@2x" width="734" src="https://user-images.githubusercontent.com/40808343/265231980-5f8187f7-93b0-45ad-ad82-02f068d1c2b1.png">
> 
> I think next steps for further developing the driver would be:
> 
>     * Testing and debugging with a wired Xbox 360 controller and getting it work in the first place.
> 
>     * Implementing the proper support for XboxOne and XboxOriginal controllers (stubs for XboxOne controller support are already in the repo)
> 
>     * Maybe building a UI / updating the UI over from the original 360Controller driver. The old UI should work the same with the new driver without any meaningful updates I think. But it might be a good idea to port the UI over from being a prefpane to being a normal app, because prefpanes are really hard to develop under modern macOS.
> 
>     * Maybe add the remapping and deadzone customization features like the original 360Controller driver, so everything in the UI works. Shouldn't be too much work.
> 
>     * Implementing support for peripherals like the microphone if anyone wants to.
> 
>     * Fixing all the memory leaks and bad error handling in my horrible code
> 
> 
> If you have any more questions, let me know.

any news?

**Comment by noah-nuebling** [comment]
@jonasw8 Nope, unfortunately I won't spend time on this any time soon, because I want to focus on my project Mac Mouse Fix. 

I'm not sure much will come from this even if somebody spends a lot of time since it seems that you need permission from Apple for each controller you want to support and they denied it to me for the controller I have.

Hopefully the community will figure something out, but it's not looking great :/

**Comment by kailous** [comment]
Your effort to port the 360Controller to DriverKit is seriously awesome – you're a hero for taking this on.

Have you thought about reaching out to third-party game controller manufacturers? Although the macOS gaming market isn't as vast as Windows, they might still be interested in supporting macOS. Your project could catch their attention, perhaps opening up opportunities for support or collaboration.

Keep it up! I'm really looking forward to the progress of your project; my third-party controller has been collecting dust for far too long.

**Comment by jonasw8** [comment]
Hi Guys! Breaking News

Sequoia received support for new joysticks

-     Wired Xbox 360
-     Nintendo Switch N64 Controller
-     Nintendo JoyCon Grip
-     Wired Xbox ONE

Would it be possible to add support for wireless xbox360?

**Comment by kokroo** [comment]
@noah-nuebling You might want to join us at https://github.com/WiiController/VirtualController/issues/1

We are continuing the goal of your project in a way, and your experience with Apple's entitlement process and your insight into the inner workings of Xbox 360 controllers will be useful and very much appreciated!

**Comment by noah-nuebling** [comment]
@kokroo thank you for your invitation! I don't think I have much to offer in this space aside from the information you can find in this thread and in the [the repo](https://github.com/noah-nuebling/mac-gamepad-driver). However, if you have any specific questions that you think I might be able to help you with feel free to reach out!
